src/zzip.cpp

```diff
@@ -0,0 +1,882 @@
+    }
+    return { base, entry_len };
+}
+
+} // namespace
+
+bool zzip::add_file( const std::filesystem::path &zzip_relative_path, std::string_view content )
+{
+    size_t estimated_size = ZSTD_compressBound( content.length() );
+
+    JsonObject footer_copy = copy_footer();
+    footer_copy.allow_omitted_members();
+    zzip_footer footer{ footer_copy };
+
+    std::optional<zzip_meta> meta_opt = footer.get_meta();
+    size_t old_content_end = 0;
+    if( meta_opt.has_value() ) {
+        old_content_end = meta_opt->content_end;
+    }
+
+    std::string relative_path_string = zzip_relative_path.generic_u8string();
+    // Add 1024 bytes padding for the footer. Should be very plenty.
+    // Skippable frames are 8 bytes of header plus the content
+    if( !ensure_capacity_for(
+            old_content_end + ZSTD_SKIPPABLEHEADERSIZE + relative_path_string.length() + kChecksumFrameSize +
+            estimated_size + 1024 ) ) {
+        return false;
+    }
+
+    size_t final_size = write_file_at( relative_path_string, content, old_content_end );
+
+    if( final_size == 0 ) {
+        return false;
+    }
+
+    std::vector<compressed_entry> new_entry;
+    new_entry.emplace_back( zzip::compressed_entry{
+        std::move( relative_path_string ),
+        old_content_end,
+        final_size
+    } );
+
+    if( !update_footer( footer_copy, old_content_end + final_size, new_entry ) ) {
+        return false;
+    }
+
+    return true;
+}
+
+bool zzip::has_file( const std::filesystem::path &zzip_relative_path ) const
+{
+    zzip_footer footer{ footer_ };
+    std::optional<zzip_file_entry> fparams = footer.get_entry( zzip_relative_path );
+    return fparams.has_value();
+}
+
+size_t zzip::get_file_size( const std::filesystem::path &zzip_relative_path ) const
+{
+    zzip_footer footer{ footer_ };
+    std::optional<zzip_file_entry> fparams = footer.get_entry( zzip_relative_path );
+    if( !fparams.has_value() ) {
+        return 0;
+    }
+    void *base = file_base_plus( fparams->offset );
+    size_t len = fparams->len;
+    std::tie( base, len ) = skip_header_frames( base, len );
+    return ZSTD_decompressBound( base, len );
+}
+
+std::vector<std::byte> zzip::get_file( const std::filesystem::path &zzip_relative_path ) const
+{
+    std::vector<std::byte> buf{};
+    size_t size = get_file_size( zzip_relative_path );
+    buf.resize( size );
+    get_file_to( zzip_relative_path, buf.data(), size );
+    return buf;
+}
+
+size_t zzip::get_file_to( const std::filesystem::path &zzip_relative_path, std::byte *dest,
+                          size_t dest_len ) const
+{
+    zzip_footer footer{ footer_ };
+    std::optional<zzip_file_entry> fparams = footer.get_entry( zzip_relative_path );
+    if( !fparams.has_value() ) {
+        return 0;
+    }
+    size_t file_len = fparams->len;
+    void *file_base = file_base_plus( fparams->offset );
+    // TODO should we checksum the file before attempting decompressing?
```

Comment:
Flagging a TODO, in case you intended to address it but missed it.
(I don't have a strong opinion on whether or not it's a blocker for the PR, but I doubt anyone except you would touch this part after it lands.
To answer the stated question - yes, we should. Otherwise there's little point for the checksums being there in the first place)

Reply:
Yeah that was a vestigial comment. Thanks!

Reply:
Actually imo, the checksums are there for detecting corruption in the frame in the case of a power loss event. xxh3 is not super fast. I could benchmark what it would add to access times to always checksum though.

Reply:
The xxh3 [marketing materials](https://github.com/Cyan4973/xxHash) say that it is, in fact, super fast

Reply:
I mean, I guess. We're not using the asm accelerated versions I don't think. But bench first and complain later I suppose.